### PR TITLE
executor: fix the issue that `pb` memory cannot be released quickly

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Fix the issue that `pb` memory cannot be released quickly.

### What is changed and how it works?
In this [PR](#10236)(stmtctx.go:296), we let `sc` hold `detail` until the query finish to get more coptask's execution information.

Because `detail` is nested in `copResponse`, if we hold its pointer, the Golang GC scanner will regard `copResponse` as active, and will not release until the query finish.

Then another field `pbResp` held by `copResonse` will not be released quickly.

Since `pbResp` is used to buffer `pb` data, it results in more memory consumption.

### Check List 

Tests

### CODE REVIEW

**Comment by qw4990** [comment]
We need more tests for this case.
Do you have any suggestions about how to test this case?
@zz-jason @XuHuaiyu @lysu

**Comment by zz-jason** [comment]
> We need more tests for this case.
> Do you have any suggestions about how to test this case?
> @zz-jason @XuHuaiyu @lysu

Can we check the two numbers of `runtime.MemStats.HeapObjects` before and after calling `runtime.GC()`?

**Comment by qw4990** [comment]
> > We need more tests for this case.
> > Do you have any suggestions about how to test this case?
> > @zz-jason @XuHuaiyu @lysu
> 
> Can we check the two numbers of `runtime.MemStats.HeapObjects` before and after calling `runtime.GC()`?

There are many unit tests running parallelly, so they may influence each other to let this unit test unstable.

**Comment by lysu** [comment]
emm..parallel running is a problem...

maybe we can make an `inuse assert util`...

the idea is that after running just fetch a heap profile, using `runtime.MemProfile`, just stealing code from 

https://github.com/golang/go/blob/cfbe3cfbeb6b2de561fc709b8644d4d7a4e182bb/src/runtime/pprof/pprof.go#L533

then just loop `[]runtime.MemProfileRecord`

https://github.com/golang/go/blob/f9a4ae018d99c5afb0e4f128545ff26e01d7b498/src/runtime/pprof/protomem.go#L38

then if we can find any MemProfileRecord that:

- stack contains current running test method name e.g. `TestDistSqlLeak`
- locate line meet some condition?
- and inuse > 0?

**Comment by lysu** [comment]
....but this case held memory is alloc in another goroutine...stack isn't start with TestDistSqlLeak :rofl:

**Comment by qw4990** [comment]
After discussing with @XuHuaiyu and reviewing [`check.RunAll`](https://github.com/pingcap/check/blob/master/run.go#L100), we can use [`SerialSuites `](https://github.com/pingcap/check/blob/master/run.go#L33) to let our UT be run serially.
Using `SerialSuites` and `runtime.GC()` can meet the test requirement.

**Comment by qw4990** [comment]
A unit test has been added, PTAL~ @XuHuaiyu @lysu @zz-jason

File: executor/memory_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	c.Assert(err, IsNil)
+	_, err = se.Execute(context.Background(), "create database test_mem")
+	c.Assert(err, IsNil)
+	_, err = se.Execute(context.Background(), "use test_mem")
+	c.Assert(err, IsNil)
+
+	// prepare data
+	totalSize := uint64(256 << 20) // 256MB
+	blockSize := uint64(8 << 10)   // 8KB
+	delta := totalSize / 5
```
Comment (XuHuaiyu) [inline executor/memory_test.go]: What does this mean?

Comment (qw4990) [reply executor/memory_test.go]: Just an empirical value, used to judge if memory leak happens.
The difference between `InUseHeap` before and after a query must be less than it.

File: executor/memory_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			mockstore.WithMVCCStore(s.mvccStore),
+		)
+		c.Assert(err, IsNil)
+		s.store = store
+		session.SetSchemaLease(0)
+		session.SetStatsLease(0)
+	}
+	d, err := session.BootstrapSession(s.store)
+	c.Assert(err, IsNil)
+	d.SetStatsUpdating(true)
```
Comment (zz-jason) [inline executor/memory_test.go]: no need to perform?

File: executor/memory_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	// check memory before close
+	runtime.GC()
+	allocatedAfter, inUseAfter := s.readMem()
+	c.Assert(allocatedAfter-allocatedBegin, GreaterEqual, totalSize)
+	c.Assert(s.memDiff(inUseAfter, inUseBegin), Less, delta)
+
+	se.Close()
+	runtime.GC()
+	allocatedFinal, inUseFinal := s.readMem()
+	c.Assert(allocatedFinal-allocatedAfter, Less, delta)
```
Comment (XuHuaiyu) [inline executor/memory_test.go]: Why do we check `allocatedFinal-allocatedAfter` here?

Comment (qw4990) [reply executor/memory_test.go]: To ensure that these executors don't allocate much memory when closing.

File: executor/memory_test.go (hunk 2 above)
Comment (qw4990) [reply executor/memory_test.go]: Some unnecessary code in `testMemoryLeak` has been removed now.

File: executor/memory_test.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+			break
+		}
+	}
+	c.Assert(rowCnt, Equals, int(numRows))
+
+	// check memory before close
+	runtime.GC()
+	allocatedAfter, inUseAfter := s.readMem()
+	c.Assert(allocatedAfter-allocatedBegin, GreaterEqual, totalSize)
+	c.Assert(s.memDiff(inUseAfter, inUseBegin), Less, delta)
```
Comment (XuHuaiyu) [inline executor/memory_test.go]: Do we need to check `s.memDiff(inUseAfter, inUseBegin)` before line 86?

Comment (qw4990) [reply executor/memory_test.go]: This check ensures that the memory used in this query can be recycled by the GC scanner.
If we put this check before line 86 and GC doesn't happen when processing this query, the value of `s.memDiff(inUseAfter, inUseBegin)` would be indeterminate.

**Pull request merged**